LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ExclusiveProcessor.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+#pragma warning disable CA1003 // Use generic event handler instances (suppressed for performance)
+
+namespace LoRaWan.NetworkServer
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    public sealed class ExclusiveProcessor<T>
```

Comment:
Is it possible that this does not have any unit tests?

Reply:
You are right, it doesn't. I feel this is general/generic _infrastructure code_ that supports `SynchronizedLoRaDeviceClient`. It could have existed directly as part of the client code but was extracted into a separate class since it's _technically_ unrelated and could be explored and reasoned about separately. Since `SynchronizedLoRaDeviceClient` uses this type, it is indirectly getting tested. I wanted to avoid duplicating the tests. According to VS code coverage, 90% of the class is exercised:

| Code                                                    | Covered (% Blocks) |
|---------------------------------------------------------|:------------------:|
| `ExclusiveProcessor.<>c<T>`                             |       100.00%      |
| `ExclusiveProcessor.<ProcessAsync>d__22<T, TResult>`    |       100.00%      |
| `ExclusiveProcessor.<TryProcessAsync>d__23<T, TResult>` |       78.33%       |
| `ExclusiveProcessor.ProcessingOutcome<T, TResult>`      |       66.67%       |
| `ExclusiveProcessor.ProcessingTaskOutcome<T, TResult>`  |       100.00%      |
| `ExclusiveProcessor<T>`                                 |       100.00%      |

Some of the untested parts are due to argument validation and never hitting the case of _interrupted processes_ since we don't take advantage of custom scheduling where one process could be given a higher priority over another; the default behaviour being FIFO:

```c#
public ExclusiveProcessor() : this(ps => ps[0]) { }
```

If you feel strongly about covering this then I can add the unit tests or we can defer that decision to when we see a drastic fall in the coverage of this class (at which point I would even question how much of it is really being used).

Reply:
As discussed - we should add unit tests for this. Also in light of the re-usability and it is a public class. Resolve this when you opened the item to track it @atifaziz